User defined window functions blog post

Adez017 opened the pull request:
## Is your feature request related to a problem or challenge?

- Closes https://github.com/apache/datafusion/issues/6781



## Describe the solution you'd like
A Blog that contains :
introduce window functions in SQL
Introduce siding windows
why computing them is hard
how to do it in DataFusion, and show how fast it can be

Adez017 commented:
hi @alamb ,  i think there is some issue while running it locally from docker image , just  raising a PR , I think it pretty nice ,  if you think  any modification needed please let me know

Dandandan commented on content/blog/2025-04-04-datafusion-userdefined-window-functions.md:
```diff
@@ -0,0 +1,154 @@
+
+With the addition of sliding window support and user-defined aggregates, DataFusion continues its march toward being a high-performance analytical engine that balances power, extensibility, and speed.
+
+Window functions may be common in SQL, but *efficient and extensibl*e window engines are rare — and now DataFusion is one of them.
```
```suggestion
Window functions may be common in SQL, but *efficient and extensible* window engines are rare — and now DataFusion is one of them.
```

Dandandan commented on content/blog/2025-04-04-datafusion-userdefined-window-functions.md:
````diff
@@ -0,0 +1,154 @@
+
+```
+DataFusion outperforms traditional SQL engines by leveraging [Apache Arrow](https://arrow.apache.org/) optimizations, making it a great choice for analytical workloads .
+Note: The reference has been taken from [@andygrove]'s blog . [see](https://andygrove.io/2019/04/datafusion-0.13.0-benchmarks/)
````
I don't see these numbers?

Adez017 replied:
![image](https://github.com/user-attachments/assets/e66b4e47-8864-4b73-bd1c-f481a2e93085)

Yes @Dandandan  you are right !. Actually i took the reference from here but if  had any better suggestions or stats i would welcome you to please help with that . once again thank you for your consideration

alamb commented on content/blog/2025-04-04-datafusion-userdefined-window-functions.md:
```diff
@@ -0,0 +1,154 @@
+
+Writing a user defined window function is slightly more complex than an aggregate function due
+to the variety of ways that window functions are called. I recommend reviewing the
+[online documentation](https://datafusion.apache.org/library-user-guide/adding-udfs.html)
```
It might help to link to the subsection https://datafusion.apache.org/library-user-guide/adding-udfs.html#registering-a-window-udf

Adez017 replied:
Done 👍

alamb commented on content/blog/2025-04-04-datafusion-userdefined-window-functions.md:
```diff
@@ -0,0 +1,154 @@
+to the variety of ways that window functions are called. I recommend reviewing the
+[online documentation](https://datafusion.apache.org/library-user-guide/adding-udfs.html)
+for a description of which functions need to be implemented. The details of how to implement
+these generally follow the same patterns as described above for aggregate functions.
```
I don't see any description above for aggregate functions 🤔

Adez017 replied:
I had updated the link you can now review it ! @alamb

alamb commented on content/blog/2025-04-04-datafusion-userdefined-window-functions.md:
```diff
@@ -0,0 +1,154 @@
+
+## Why Computing Sliding Windows Is Hard
+
+Imagine you’re at a café, and the barista is preparing coffee orders. If they made each cup from scratch without using pre-prepared ingredients, the process would be painfully slow. This is exactly the problem with naïve sliding window computations.
```
this is a pretty cool analogy

Adez017 replied:
thank you

alamb commented on content/blog/2025-04-04-datafusion-userdefined-window-functions.md:
```diff
@@ -0,0 +1,154 @@
+
+With the addition of sliding window support and user-defined aggregates, DataFusion continues its march toward being a high-performance analytical engine that balances power, extensibility, and speed.
+
+Window functions may be common in SQL, but *efficient and extensible* window engines are rare — and now DataFusion is one of them.
```
this might be a good sentence to move up

Adez017 replied:
Sure !

alamb commented on content/blog/2025-04-04-datafusion-userdefined-window-functions.md:
```diff
@@ -0,0 +1,154 @@
+
+## Final Thoughts and Recommendations 
+
+With the addition of sliding window support and user-defined aggregates, DataFusion continues its march toward being a high-performance analytical engine that balances power, extensibility, and speed.
```
Given window functions were added a while ago we may want to emphasize that DataFUsion has had window functions since XXX (look up when they were added)

Adez017 replied:
I am not able to find the dates when this got released it would be great help

alamb commented on content/blog/2025-04-04-datafusion-userdefined-window-functions.md:
```diff
@@ -0,0 +1,154 @@
+This helps in analytical queries where we need cumulative sums, moving averages, or ranking without losing individual records.
+
+
+## User Defined Window Functions
```
I think DataFusion is somewhat rare in the database world for supporting user defined window functions

Perhaps we can add some background here such as "while many databases support user defined scalar and user defined aggregate functions, user defined window functions are not as common" (ideally we could find some evidence to back up that claim too)

Adez017 replied:
Sure !, i would start working on that

alamb commented on content/blog/2025-04-04-datafusion-userdefined-window-functions.md:
```diff
@@ -0,0 +1,154 @@
+
+Many traditional query engines struggle to optimize these computations effectively, leading to sluggish performance.
+
+## How DataFusion Making it fast
```
```suggestion
## How DataFusion Evaluates Window Functions Quickly
```

alamb commented on content/blog/2025-04-04-datafusion-userdefined-window-functions.md:
````diff
@@ -0,0 +1,154 @@
+  ) AS moving_score
+FROM leaderboard;
+```
+This gives you full flexibility to build **domain-specific logic** that plugs seamlessly into DataFusion’s engine — all without sacrificing performance.
````
It would be great to expand this section a little more with what a user defined aggregate looks like -- perhaps copying soem/all of the example from https://datafusion.apache.org/library-user-guide/adding-udfs.html#registering-a-window-udf

Adez017 replied:
👍

Adez017 replied:
done that

alamb commented on content/blog/2025-04-04-datafusion-userdefined-window-functions.md:
```diff
@@ -0,0 +1,154 @@
+
+## Performance Gains
+
+To demonstrate efficiency, we benchmarked a 1-million row dataset with a sliding window aggregate.
```
If we are going to report numbers like this I think it is important to be clear what query was run on what setup (so they can be reproduced if desired)

Adez017 replied:
what can be ideal approach ? . it would be great if anyone wants to help with this specific section @alamb  @Dandandan

alamb replied:
I suggest we avoid comparisons with other systems and simply focus on DataFusion's own performance.

Adez017 replied:
so do i need to remove this section ?

Adez017 replied:
okay , i also feel it quite so

alamb reviewed (commented):
Thank you so much @Adez017  -- this is looking great

Given the title of this post is about user defined window functions, there is not very much discussion about that topic -- most of the content is about window functions in general

Perhaps you can structure the post so that it
1. Introduces window functions in general and what they do (you have most of this content already)
2. Walks through an example of creating a user defined window functions in a bit more detail (I think you can borrow / copy liberally from https://datafusion.apache.org/library-user-guide/adding-udfs.html#registering-a-window-udf)

What do you think?

And thanks again for starting this post

Adez017 commented:
> Thank you so much @Adez017 -- this is looking great
> 
> Given the title of this post is about user defined window functions, there is not very much discussion about that topic -- most of the content is about window functions in general
> 
> Perhaps you can structure the post so that it
> 
> 1. Introduces window functions in general and what they do (you have most of this content already)
> 2. Walks through an example of creating a user defined window functions in a bit more detail (I think you can borrow / copy liberally from https://datafusion.apache.org/library-user-guide/adding-udfs.html#registering-a-window-udf)
> 
> What do you think?
> 
> And thanks again for starting this post

I had done a bunch of updates regarding your advice . i need you to review it again and give the feedback . Thank you very much for the efforts

Adez017 commented:
any updates ?

alamb commented:
Thanks @Adez017  -- I very much have this on my list of things to review. I expect to do so later this afternoon

Adez017 commented:
> Thanks @Adez017 -- I very much have this on my list of things to review. I expect to do so later this afternoon

Great to hear @alamb  . i am no sure about the `Performance Gains` section . we can have to have a closer look on that

alamb commented on content/blog/2025-04-04-datafusion-userdefined-window-functions.md:
````diff
@@ -0,0 +1,339 @@
+       AVG(sales) OVER (ORDER BY date ROWS BETWEEN 6 PRECEDING AND CURRENT ROW) AS moving_avg
+FROM sales;
+```
+Here, each row’s result is computed based on the last 7 days, making it computationally intensive as data grows.
````
I am not sure I would say a 7 day window is computationally expensive -- if you changed the example to be `UNBOUNDED PRECEDING` you could make the point that the window size kept growing. Or 365 days maybe would make the point better?

alamb commented on content/blog/2025-04-04-datafusion-userdefined-window-functions.md:
```diff
@@ -0,0 +1,339 @@
+
+Many traditional query engines struggle to optimize these computations effectively, leading to sluggish performance.
+
+## How DataFusion Evaluates Window Functions Quickly
+In the world of big data, every millisecond counts. Imagine you’re analyzing stock market data, tracking sensor readings from millions of IoT devices, or crunching through massive customer logs—speed matters. This is where [DataFusion](https://datafusion.apache.org/) shines, making window function computations blazing fast. Let’s break down how it achieves this remarkable performance.
```
This feels a little over dramatic -- perhaps we can tone down the wording some

In general I don't think DataFusion's window function performance is significantly better (or worse) than other engines . I think the main difference is how extensible it is. I'll try and comment more inline

alamb commented on content/blog/2025-04-04-datafusion-userdefined-window-functions.md:
```diff
@@ -0,0 +1,339 @@
+    }
+}
+
+/// Different evaluation methods are called depending on the various
+/// settings of WindowUDF. This example uses the simplest and most
+/// general, `evaluate`. See `PartitionEvaluator` for the other more
+/// advanced uses.
+impl PartitionEvaluator for MyPartitionEvaluator {
```
In this blog form, it might be easier to read if you broke the comments out of the text so something lik e

```suggestion
```(end text)
Different evaluation methods are called depending on the various
settings of WindowUDF. In the first example, we  use the simplest and most
general, `evaluate`. we will seee how to use `PartitionEvaluator` for the other more
 advanced uses later in the article.
 
 ```rust
impl PartitionEvaluator for MyPartitionEvaluator {
```

alamb commented on content/blog/2025-04-04-datafusion-userdefined-window-functions.md:
````diff
@@ -0,0 +1,339 @@
+}
+```
+### Registering a Window UDF
+To register a Window UDF, you need to wrap the function implementation in a `WindowUDF` struct and then register it with the `SessionContext`. DataFusion provides the `create_udwf` helper functions to make this easier. There is a lower level API with more functionality but is more complex, that is documented in [advanced_udwf.rs](https://github.com/apache/datafusion/blob/main/datafusion-examples/ ...
````
It would be nice if you could add links to the doc for `WindowUDF`, `SessionContext`, etc here

Adez017 replied:
got it

alamb reviewed (commented):
Thanks again @Adez017  -- I took another pass over this post and left some more comments -- let me know what you think

alamb commented on content/blog/2024-11-19-datafusion-python-udf-comparisons.md:
```diff
@@ -137,7 +137,7 @@ def is_of_interest_impl(
         suppkey = suppkey_arr[idx].as_py()
         returnflag = returnflag_arr[idx].as_py()
-        value = (partkey, suppkey, returnflag)
+        value = (partkey, suppkey, returnflag)  
```
This seems to have been changed by accident: 

```suggestion
        value = (partkey, suppkey, returnflag)
```

Adez017 commented on content/blog/2025-04-04-datafusion-userdefined-window-functions.md:
```diff
@@ -0,0 +1,339 @@
+
+In this post, we'll explore:
+
+- What window functions are and why they matter
```
will do that

comphead commented on content/blog/2025-04-17-user-defined-window-functions.md:
````diff
@@ -0,0 +1,427 @@
+For instance, if we want a 7-day moving average of sales:
+
+```sql
+SELECT date, sales, 
````
would be good to show query results here

Adez017 replied:
yeah Sure!

comphead commented on content/blog/2025-04-17-user-defined-window-functions.md:
````diff
@@ -0,0 +1,427 @@
+```sql
+SELECT 
+  FIRST_VALUE(price) OVER (PARTITION BY date_bin('1 month', time) ORDER BY time DESC), 
+  FIRST_VALUE(price) OVER (PARTITION BY date_bin('1 month', time) ORDER BY time DESC)
````
is it expected to have two same functions? do you mean to calc last_value?

comphead commented on content/blog/2025-04-17-user-defined-window-functions.md:
```diff
@@ -0,0 +1,427 @@
+SELECT 
+  FIRST_VALUE(price) OVER (PARTITION BY date_bin('1 month', time) ORDER BY time DESC), 
+  FIRST_VALUE(price) OVER (PARTITION BY date_bin('1 month', time) ORDER BY time DESC)
+  AVG(price)         OVER (PARTITION BY date_bin('1 month', time)),
```
```suggestion
  AVG(price)         OVER (PARTITION BY date_bin('1 month', time))
```

comphead commented on content/blog/2025-04-17-user-defined-window-functions.md:
```diff
@@ -0,0 +1,427 @@
+        // point numbers to calculate a moving average
+        let arr: &Float64Array = values[0].as_ref().as_primitive::<Float64Type>();
+
+        let range_len = range.end - range.start;
```
in the example it is better to comment every line

comphead commented on content/blog/2025-04-17-user-defined-window-functions.md:
````diff
@@ -0,0 +1,427 @@
+### Registering a Window UDF
+To register a Window UDF, you need to wrap the function implementation in a `WindowUDF` struct and then register it with the `SessionContext`. DataFusion provides the `create_udwf` helper functions to make this easier. There is a lower level API with more functionality but is more complex, that is documented in [advanced_udwf.rs](https://github.com/apache/datafusion/blob/main/datafusion-examples/ ...
+
+```sql
````
```suggestion
```rust
```

comphead commented on content/blog/2025-04-17-user-defined-window-functions.md:
```diff
@@ -0,0 +1,427 @@
+// here is where we define the UDWF. We also declare its signature:
+let smooth_it = create_udwf(
+    "smooth_it",
+    DataType::Float64,
```
please comment params it would be easier to read

comphead commented on content/blog/2025-04-17-user-defined-window-functions.md:
```diff
@@ -0,0 +1,427 @@
+
+A heartfelt thank you to [@alamb] and [@andygrove] for their invaluable reviews and thoughtful feedback—they’ve been instrumental in shaping this post.
+
+The Apache Arrow and DataFusion communities are vibrant, welcoming, and full of passionate developers building something truly powerful. If you’re excited about high-performance analytics and want to be part of an open-source journey, I highly encourage you to explore the [official documentation]((https://datafusion.apache.org/)) and dive into one of the many [open issues](https://github.com/apac ...
```
```suggestion
The Apache Arrow and Apache DataFusion communities are vibrant, welcoming, and full of passionate developers building something truly powerful. If you’re excited about high-performance analytics and want to be part of an open-source journey, I highly encourage you to explore the [official documentation]((https://datafusion.apache.org/)) and dive into one of the many [open issues](https://github.com/apache/datafusion/issues). There’s never been a better time to get involved!
```

Adez017 replied:
done 👍

getChan commented on content/blog/2025-04-17-user-defined-window-functions.md:
````diff
@@ -0,0 +1,427 @@
+```sql
+SELECT id, value, SUM(value) OVER (ORDER BY id) AS running_total
+FROM sales;
+```
````
Why did you put the query in here? It seems unrelated to the context.

Adez017 commented on content/blog/2025-04-04-datafusion-userdefined-window-functions.md:
```diff
@@ -0,0 +1,346 @@
+{% endcomment %}
+-->
+
+## Introduction
```
will do it now !

Adez017 commented:
> Thank you very much @Adez017
> 
> I took another pass through this post and did
> 
> 1. some small wordsmithing and formatting tweaks.
> 2. Added a section on how window functions are implemented
> 3. Updated the date
> 
> I just pushed the commits directly to your branch to hopefully reduce the back and forth
> 
> I think this blog is now ready to publish, but it would be great if you had a chance to review it again. I will also send a note to the mailing list and slack/discord,

Hi @alamb i would love to do so , get back to you soon as  i finish reviewing it

Adez017 commented on content/blog/2025-04-17-user-defined-window-functions.md:
```diff
@@ -0,0 +1,427 @@
+
+## Why Computing Sliding Windows Is Hard
+
+Imagine you’re at a café, and the barista is preparing coffee orders. If they made each cup from scratch without using pre-prepared ingredients, the process would be painfully slow. This is exactly the problem with naïve sliding window computations.
```
Yes beacuse it is computationally hard , same as we mentioned in our example

Adez017 commented:
hi @alamb  , i had done some changes about spells , etc . i think it is ready to be merged .  i would prefer please look at the post one final time before merging . for me , its looks pretty good and awsome , upto the mark .😁

alamb commented:
Thanks @Adez017  ! Let's give it another day or two for any remaining comments and then I'll plan to publish it

Adez017 commented:
Sure @alamb  👍

Adez017 commented:
> Thanks @Adez017 ! Let's give it another day or two for any remaining comments and then I'll plan to publish it

I think now , we should move forward

alamb commented:
Thanks again @Adez017  @Dandandan @getChan  and @comphead

alamb commented:
The blog is now live: https://datafusion.apache.org/blog/2025/04/19/user-defined-window-functions/
